Public/Src/Utilities/Instrumentation/Tracing/Log.cs

```diff
@@ -108,13 +108,55 @@ public abstract partial class Logger
             Message = "Managed memory: {0} MB, private bytes: {1} MB, threads: {2}")]
         public abstract void Memory(LoggingContext context, long managedHeapMegabytes, long privateMemorySizeMegabytes, int threads);
 
+        /// <summary>
+        /// Logs an UnexpectedCondition both locally as a warning and a capped number of times per build session to telemetry.
+        /// Always call this wrapper function
+        /// </summary>
+        /// <param name="loggingContext"></param>
+        /// <param name="description"></param>
+        public void UnexpectedCondition(LoggingContext loggingContext, string description)
+        {
+            lock (s_currentSessionId)
+            {
+                // Check and reset the context for what build session this in case the server process is reused
+                if (s_currentSessionId != loggingContext.GetRootContext().Session.Id)
+                {
+                    s_currentSessionId = loggingContext.GetRootContext().Session.Id;
+                    s_unexpectedConditionsLogged = 0;
+                }
+
+                s_unexpectedConditionsLogged++;

```

Comment:
>s_unexpectedConditionsLogged++; [](start = 16, length = 31)

I know it's highly unlikely, but what do you think about adding some check against an overflow? maybe move the increment inside `if (...) {}` ?

Reply:
Good idea. We should expect the unexpected for what is unexpected